client/extensions/woocommerce/woocommerce-services/state/shipping-label/reducer.js

```diff
@@ -686,14 +695,202 @@ reducers[ WOOCOMMERCE_SERVICES_SHIPPING_LABEL_SAVE_PACKAGES ] = state => {
 	};
 };
 
-reducers.WCS_TOGGLE_DUMMY_FIELD = ( state, { value } ) => {
+reducers[ WOOCOMMERCE_SERVICES_SHIPPING_LABEL_SET_CONTENTS_TYPE ] = (
+	state,
+	{ packageId, value }
+) => {
+	return {
+		...state,
+		form: {
+			...state.form,
+			packages: {
+				...state.form.packages,
+				selected: {
+					...state.form.packages.selected,
+					[ packageId ]: {
+						...state.form.packages.selected[ packageId ],
+						contentsType: value,
+						contentsExplanation: '',
+					},
+				},
+			},
+		},
+	};
+};
+
+reducers[ WOOCOMMERCE_SERVICES_SHIPPING_LABEL_SET_CONTENTS_EXPLANATION ] = (
+	state,
+	{ packageId, value }
+) => {
+	return {
+		...state,
+		form: {
+			...state.form,
+			packages: {
+				...state.form.packages,
+				selected: {
+					...state.form.packages.selected,
+					[ packageId ]: {
+						...state.form.packages.selected[ packageId ],
+						contentsExplanation: value,
+					},
+				},
+			},
+		},
+	};
+};
+
+reducers[ WOOCOMMERCE_SERVICES_SHIPPING_LABEL_SET_RESTRICTION_TYPE ] = (
+	state,
+	{ packageId, value }
+) => {
+	return {
+		...state,
+		form: {
+			...state.form,
+			packages: {
+				...state.form.packages,
+				selected: {
+					...state.form.packages.selected,
+					[ packageId ]: {
+						...state.form.packages.selected[ packageId ],
+						restrictionType: value,
+						restrictionExplanation: '',
+					},
+				},
+			},
+		},
+	};
+};
+
+reducers[ WOOCOMMERCE_SERVICES_SHIPPING_LABEL_SET_RESTRICTION_EXPLANATION ] = (
+	state,
+	{ packageId, value }
+) => {
+	return {
+		...state,
+		form: {
+			...state.form,
+			packages: {
+				...state.form.packages,
+				selected: {
+					...state.form.packages.selected,
+					[ packageId ]: {
+						...state.form.packages.selected[ packageId ],
+						restrictionExplanation: value,
+					},
+				},
+			},
+		},
+	};
+};
+
+reducers[ WOOCOMMERCE_SERVICES_SHIPPING_LABEL_SET_ABANDON_ON_NON_DELIVERY ] = (
+	state,
+	{ packageId, value }
+) => {
+	return {
+		...state,
+		form: {
+			...state.form,
+			packages: {
+				...state.form.packages,
+				selected: {
+					...state.form.packages.selected,
+					[ packageId ]: {
+						...state.form.packages.selected[ packageId ],
+						abandonOnNonDelivery: Boolean( value ),
+					},
+				},
+			},
+		},
+	};
+};
```

Comment:
holy toledo batman! that's a lot of nesting and cloning 😱 and all for a single boolean!

I don't have any blocker here because I sense that what I would like to write has broad implications. nonetheless it would be good for us to be thinking about flat state structure. right now the selected packages are stored deep in the nested data hierarchy and changing one requires changing everything up the links. we have options for flat hierarchies where it's not so hard to change.

also, is this not synchronizing with the server somewhere? it would be nice if we could push out an update to the server and let it provide the update as an atomic replace operation here rather than have all these huge functions whose purpose are to flip a single bit.

Reply:
> holy toledo batman! that's a lot of nesting and cloning 😱 and all for a single boolean!

I know... but it's The Redux Way ™️ 

> Nonetheless it would be good for us to be thinking about flat state structure.

I'll keep this in mind for the future, but this would imply changes too big to tackle in this already massive PR :/
We have been modelling the state tree to keep a similar structure as the UI, but I get that it's not the most performant way of approaching it. Flatter would be better, for sure.

> is this not synchronizing with the server somewhere?

It's not synchronizing to the server. All this is local UI-related state, which at the end is smooshed into a big `POST` request to get rates for this order ([here](https://github.com/Automattic/wp-calypso/pull/24974/files/f1089ba50b743fc54c14f00e77b85439a3e24731#diff-09d52b7e14fcd8c2a349cabb5e073666R251)) and then another `POST` request to actually purchase the shipping label(s) [here](https://github.com/Automattic/wp-calypso/pull/24974/files/f1089ba50b743fc54c14f00e77b85439a3e24731#diff-09d52b7e14fcd8c2a349cabb5e073666R962).

Reply:
> I know... but it's The Redux Way ™️

Actually it isn't :wink: There's no prescribed Redux way. What I see here is a common pattern though that people use but I think misses the comparability aspect which can make it shine. It's asking the question "what data does this change?" instead of the question "what changes this data?" because Redux operates very well when one single data atom governs itself.

that is, imagine a flat structure where we have a reducer that holds a flag…

```js
const abandonOnNonDeliver = ( state = false, { type, value } ) =>
	type === WOOCOMMERCE…ON_NON_DELIVERY
		? Boolean( value )
		: state;

…

export reducer = combineReducers( {
	…
	selectedProductsAbandonOnDeliver: keyedReducer( 'productId', abandonOnNonDeliver )
	…
} )
```

the important part here isn't the details I wrote but the idea that we can compose behaviors and shard our tree in a flat way and relate those things through ids. it looks more and more like a database but one in which every change is audited and reproducible.

> I'll keep this in mind for the future, but this would imply changes too big to tackle in this already massive PR :/

Yes and that's why I don't want it to be a blocker for you. It's also why it's so important to hit these things at the beginning before we go too far down the road.

Reply:
> Actually it isn't 😉 There's no prescribed Redux way.

True, but all the tooling and libraries around Redux assume that the `state` object is immutable, so cloning the branches that hold a value that changes is inevitable.

> that is, imagine a flat structure where we have a reducer that holds a flag…

Sure, in that example the tree structure is flatter (which is good), but you still have 2 levels of nesting/cloning going on: The `combineReducers` level and the `keyedReducer` level. Not trying to dismiss that structure, it's still an improvement, but I'm just saying that the same pattern of cloning objects to mutate them is still there, even though it's hidden to the naked eye :)

Reply:
> What I see here is a common pattern though that people use but I think misses the composability aspect which can make it shine. It's asking the question "what data does this change?" instead of the question "what changes this data?" because Redux operates very well when one single data atom governs itself.

I totally agree with this. This reducer became too big over the years, it governs a big branch of the state tree and it should have been broken down further. At the very least, one reducer for every step of the form.

Reply:
I think we can start to use `combineReducers`, without overhauling everything. For example, we can create a separate reducer for `customs`, and use `combineReducers` to combine the form state we have now with `customs`. This can pave the way for us to refactor the pieces we already have as we go along.

Reply:
Splitting the tree up by using `combineReducers ` will also help us in writing some tests more easily for this :)

Reply:
also please feel empowered to flatten out reducers and state. sometimes it may feel verbose but flattening can actually improve many different aspects of state, closing is just one of them.

typically "flattening" means two things (to me at least):
 - including more reference ids in the property name
 - building join tables

if we think of a site having posts then we can illustrate the difference

```js
state: {
  sites: {
    14: {
      posts: {
        1: {}, 2: {}, …
      }
    }
  }
}
```

^^^ that's a lot of nesting and while easy to ask "what posts does this site have" it makes some things really difficult, like, "what posts does this _user account_ have?" since we have to iterate over all of the sites.

```js
state: {
  sites: { 14: {} },
  post: {
    '14-1': {},
    '14-2': {}
  },
  sitePosts: {
    14: [ '14-1', '14-2' ]
  }
}
```

^^^ while a little more verbose we can actually discover a few things here: it's still trivial to find "what posts does this site have?" because we made the optional join table. further, we can fill up that join table with ids but still have the actual posts missing from memory! that means we can reference posts that we know exist without having to load them from the remote server.

further it's trivial to operate over the list of all posts since _as they are distinct resources with their own identity anyway_ they aren't subservient to sites.

we acknowledge these things with our selectors and can actually shuffle around the structure under the hood if and when it makes sense.